File: app/src/main/java/com/example/currencyrate/MainActivity.java
package com.example.currencyrate;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.ScaleAnimation;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        CardView btnOpenConverter = findViewById(R.id.btnOpenConverter);
        btnOpenConverter.setOnClickListener(v -> {
            animateClick(v);
            startActivity(new Intent(MainActivity.this, ConverterActivity.class));
        });

        ViewGroup gridLayout = findViewById(R.id.glCurrencies);
        for (int i = 0; i < gridLayout.getChildCount(); i++) {
            View card = gridLayout.getChildAt(i);
            card.setOnClickListener(v -> {
                animateClick(v);
                
                // Извлекаем данные из карточки для передачи в детали
                String code = "USD";
                String rate = "0.00";
                
                if (v instanceof ViewGroup) {
                    ViewGroup vg = (ViewGroup) v;
                    // В нашем layout структура: FrameLayout -> LinearLayout -> TextViews
                    // Мы можем найти TextView с кодом валюты
                    for (int j = 0; j < vg.getChildCount(); j++) {
                        View child = vg.getChildAt(j);
                        if (child instanceof ViewGroup) {
                            ViewGroup innerVg = (ViewGroup) child;
                            for (int k = 0; k < innerVg.getChildCount(); k++) {
                                View text = innerVg.getChildAt(k);
                                if (text instanceof TextView) {
                                    String content = ((TextView) text).getText().toString();
                                    if (content.contains("/")) code = content.split("/")[0];
                                    if (Character.isDigit(content.charAt(0))) rate = content;
                                }
                            }
                        }
                    }
                }

                Intent intent = new Intent(MainActivity.this, DetailsActivity.class);
                intent.putExtra("CURRENCY_CODE", code);
                intent.putExtra("CURRENCY_RATE", rate);
                startActivity(intent);
            });
        }
    }

    private void animateClick(View v) {
        ScaleAnimation anim = new ScaleAnimation(1f, 0.95f, 1f, 0.95f, 
            Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(100);
        anim.setRepeatMode(Animation.REVERSE);
        anim.setRepeatCount(1);
        v.startAnimation(anim);
    }
}

File: app/src/main/java/com/example/currencyrate/ui/MainActivity.kt
package com.example.currencyrate.ui

import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.currencyrate.data.local.AppDatabase
import com.example.currencyrate.data.remote.CbrApi
import com.example.currencyrate.data.repository.CurrencyRepository
import com.example.currencyrate.databinding.ActivityMainBinding
import com.example.currencyrate.viewmodel.MainViewModel
import kotlinx.coroutines.launch
import retrofit2.Retrofit
import retrofit2.converter.simplexml.SimpleXmlConverterFactory

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    
    private val viewModel: MainViewModel by viewModels {
        object : ViewModelProvider.Factory {
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                val database = AppDatabase.getDatabase(applicationContext)
                val retrofit = Retrofit.Builder()
                    .baseUrl("https://www.cbr.ru/")
                    .addConverterFactory(SimpleXmlConverterFactory.create())
                    .build()
                val api = retrofit.create(CbrApi::class.java)
                val repository = CurrencyRepository(api, database.currencyDao())
                return MainViewModel(repository) as T
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupRecyclerView()
        setupListeners()
        observeViewModel()
    }

    private fun setupRecyclerView() {
        val adapter = CurrencyAdapter(
            onFavoriteClick = { code, isFav -> viewModel.toggleFavorite(code, isFav) },
            onItemClick = { currency ->
                val intent = Intent(this, DetailsActivity::class.java).apply {
                    putExtra("CURRENCY_CODE", currency.code)
                    putExtra("CURRENCY_NAME", currency.name)
                    putExtra("CURRENCY_RATE", currency.rate)
                }
                startActivity(intent)
            }
        )
        binding.rvCurrencies.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            this.adapter = adapter
            alpha = 0f
        }
    }

    private fun setupListeners() {
        binding.btnOpenConverter.setOnClickListener {
            startActivity(Intent(this, ConverterActivity::class.java))
        }

        binding.fabAddCurrency.setOnClickListener {
            AddCurrencyBottomSheet().show(supportFragmentManager, "AddCurrencyBottomSheet")
        }

        binding.ivConvIcon.setOnClickListener {
            viewModel.refreshRates()
        }
    }

    private fun observeViewModel() {
        viewModel.favoriteCurrencies.observe(this) { currencies ->
            (binding.rvCurrencies.adapter as CurrencyAdapter).submitList(currencies)
        }

        viewModel.isLoading.observe(this) { isLoading ->
            binding.pbLoading.isVisible = isLoading
            if (!isLoading) {
                binding.rvCurrencies.animate()
                    .alpha(1f)
                    .setDuration(500)
                    .start()
            }
        }

        viewModel.syncStatus.observe(this) { status ->
            binding.tvSubtitle.text = status
        }

        lifecycleScope.launch {
            viewModel.updateProgress.collect { progress ->
                if (progress == 0) {
                    binding.updateProgress.setProgressCompat(0, false)
                } else {
                    binding.updateProgress.setProgressCompat(progress, true)
                }
            }
        }
    }
}
